### PR TITLE
Fix #858: Added missing TopicsApi

### LINKED ISSUE #858: Tags in Project API are deprecated
In a project I'm currently working on I have to create gitlab projects via the gitlab4j-api. Thanks a lot for the great work!!
The projects we create get topics to be better able to filter based on them. It seems that gitlab4j-api so far only supports the deprecated tags and not the new recommended topics. https://docs.gitlab.com/ee/api/projects.html#create-project

Is there any plan to implement it?

Thank you in advance,
Björn

### PR DESCRIPTION
**Pull request by NadeoWolf**
Added the missing TopicsAPI. 
Due to compatibility problems with Maven, I have only been able to test the new functions manually.

### CODE REVIEW

**Review by jmini — CHANGES_REQUESTED** [review]
I think some unit tests should be extended (I can have a look).

Thank you very much for this contribution!

File: src/main/java/org/gitlab4j/api/models/Topic.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    private Integer id;
+
+    private String name;
+
+    private String title;
+
+    private String description;
+
+    private int total_projects_count;
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/models/Topic.java:15]: Should be:
```suggestion
    private int totalProjectsCount;
```

the JSON serialization is doing the conversion to attribute name `"total_projects_count"` already.

File: src/main/java/org/gitlab4j/api/models/Topic.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    private String avatar_url;
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/models/Topic.java:17]: should be `avatarUrl`

File: src/main/java/org/gitlab4j/api/models/Topic.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    public String getDescription() {
+        return description;
+    }
+
+    public void setDescription(String description) {
+        this.description = description;
+    }
+
+    public int getTotal_projects_count() {
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/models/Topic.java:51]: should be `getTotalProjectsCount()`

File: src/main/java/org/gitlab4j/api/models/ProjectFilter.java [hunk 4]
```diff
@@ -35,6 +35,9 @@ public class ProjectFilter {
     private Date lastActivityAfter;
     private Date lastActivityBefore;
     private String repositoryStorage;
+    private Boolean imported;
+    private String topic;
+    private Integer topic_id;
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/models/ProjectFilter.java:40]: should be `topicId`

File: src/main/java/org/gitlab4j/api/models/Project.java [hunk 5]
```diff
@@ -542,19 +544,44 @@ public void setStarCount(Integer starCount) {
...
+
+    public void setTopics(List<String> topics) {
+        this.topics = topics;
+    }
+
+    public Project withTopics(List<String> topics) {
+        this.topics = topics;
+        return (this);
+    }
+
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/models/Project.java:584]: This is interesting, but to use it, we need something similar to what exists for the old `TagList` in `ProjectApi.java`:

```
            if (project.getTagList() != null && !project.getTagList().isEmpty()) {
                formData.withParam("tag_list", String.join(",", project.getTagList()));
            }
 ```

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+     * @return the list of groups viewable by the authenticated userin the specified page range
+     * @throws GitLabApiException if any exception occurs
+     */
+    public List<Topic> getTopics(int page, int perPage) throws GitLabApiException {
+        Response response = get(Response.Status.OK, getPageQueryParams(page, perPage), "topics");
+        return (response.readEntity(new GenericType<List<Topic>>() {}));
+    }
+
+    /**
+     * Get a Pager of groups. (As user: my groups, as admin: all groups)
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:51]: copy paste error?

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+     * @param itemsPerPage the number of Group instances that will be fetched per page
+     * @return the list of groups viewable by the authenticated user
+     * @throws GitLabApiException if any exception occurs
+     */
+    public Pager<Topic> getTopics(int itemsPerPage) throws GitLabApiException {
+        return (new Pager<Topic>(this, Topic.class, itemsPerPage, null, "topics"));
+    }
+
+    /**
+     * Get a Stream of groups. (As user: my groups, as admin: all groups)
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:64]: copy paste error?

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+     *
+     * @return a Stream of groups viewable by the authenticated user
+     * @throws GitLabApiException if any exception occurs
+     */
+    public Stream<Topic> getGroupsStream() throws GitLabApiException {
+        return (getTopics(getDefaultPerPage()).stream());
+    }
+
+    /**
+     * Get all details of a group.
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:76]: copy paste error?

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+    /**
+     * Get a Stream of groups. (As user: my groups, as admin: all groups)
+     *
+     * <pre><code>GitLab Endpoint: GET /topics</code></pre>
+     *
+     * @return a Stream of groups viewable by the authenticated user
+     * @throws GitLabApiException if any exception occurs
+     */
+    public Stream<Topic> getGroupsStream() throws GitLabApiException {
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:71]: copy paste error? wrong method name?

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+    /**
+     * Get all details of a group as an Optional instance.
+     *
+     * <pre><code>GitLab Endpoint: GET /topics/:id</code></pre>
+     *
+     * @param id the topic ID
+     * @return the Topic for the specified topic id as an Optional instance
+     */
+    public Optional<Topic> getOptionalGroup(Integer id) {
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:97]: copy paste error? wrong method name?

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        if(isApiVersion(GitLabApi.ApiVersion.V3)){
+            throw new GitLabApiException("Topics need api v4+");
+        }
+        delete(Response.Status.NO_CONTENT,null, "topics", id);
+    }
+
+    /**
+     * You must be an administrator to merge a source topic into a target topic. When you merge topics, you delete the source topic and move all assigned projects to the target topic.
+     *
+     * <pre><code>DELETE /topics/:id</code></pre>
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:182]: copy paste error?

**Review by jmini — CHANGES_REQUESTED** [review]
There is some work to do (see comments)

**Comment by NadeoWolf** [comment]
Im on Holiday. Will check it after 4. September.

**Comment by jmini** [comment]
> Will check it after 4. September.

No problem.
I had some fixes locally. I pushed them to your branch.

File: src/main/java/org/gitlab4j/api/models/Project.java (hunk 5 above)
Comment (NadeoWolf) [reply src/main/java/org/gitlab4j/api/models/Project.java:584]: In `model/Project.java` I have used the same structure as in `withTagList`.

In `ProjectApi.java` i added TopicList to  ` updateProject(Project project)` and `createProject(Project project, String importUrl)` in the format you copied. 

i don't see the difference you are referring to.

**Review by jmini — APPROVED** [review]
This looks good to me. 👍 

I would like to perform some manual tests before I merge.

Thank you for this contribution 🎉

File: src/main/java/org/gitlab4j/api/TopicsApi.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+     * Delete the topic's avatar for the specified topic.
+     *
+     * <pre><code>GitLab Endpoint: PUT /topics/:id</code></pre>
+     *
+     * @param id the topic in the form of an Integer
+     * @return the updated Topic instance
+     * @throws GitLabApiException if any exception occurs
+     */
+    public Topic deleteTopicAvatar(final Integer id) throws GitLabApiException {
+        Response response = putUpload(Response.Status.OK, "avatar", null,  "topics", id);
```
Comment (jmini) [inline src/main/java/org/gitlab4j/api/TopicsApi.java:160]: This is failing here:

```
org.gitlab4j.api.GitLabApiException: Missing body part entity of type "application/octet-stream".
	at org.gitlab4j.api.AbstractApi.handle(AbstractApi.java:700)
	at org.gitlab4j.api.AbstractApi.putUpload(AbstractApi.java:560)
	at org.gitlab4j.api.TopicsApi.deleteTopicAvatar(TopicsApi.java:160)
	at ...
Caused by: javax.ws.rs.ProcessingException: Missing body part entity of type "application/octet-stream".
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:312)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$invoke$0(JerseyInvocation.java:662)
	at org.glassfish.jersey.client.JerseyInvocation.call(JerseyInvocation.java:697)
	at org.glassfish.jersey.client.JerseyInvocation.lambda$runInScope$3(JerseyInvocation.java:691)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:205)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:390)
	at org.glassfish.jersey.client.JerseyInvocation.runInScope(JerseyInvocation.java:691)
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:661)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:439)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.put(JerseyInvocation.java:328)
	at org.gitlab4j.api.GitLabApiClient.putUpload(GitLabApiClient.java:696)
	at org.gitlab4j.api.GitLabApiClient.putUpload(GitLabApiClient.java:677)
	at org.gitlab4j.api.AbstractApi.putUpload(AbstractApi.java:558)
	... 12 more
Caused by: java.lang.IllegalArgumentException: Missing body part entity of type "application/octet-stream".
	at org.glassfish.jersey.media.multipart.internal.MultiPartWriter.writeTo(MultiPartWriter.java:191)
	at org.glassfish.jersey.media.multipart.internal.MultiPartWriter.writeTo(MultiPartWriter.java:56)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:242)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:227)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1116)
	at org.glassfish.jersey.client.ClientRequest.doWriteEntity(ClientRequest.java:461)
	at org.glassfish.jersey.client.ClientRequest.writeEntity(ClientRequest.java:443)
	at org.glassfish.jersey.client.internal.HttpUrlConnector._apply(HttpUrlConnector.java:400)
	at org.glassfish.jersey.client.internal.HttpUrlConnector.apply(HttpUrlConnector.java:273)
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:300)
	... 26 more
```

Comment (NadeoWolf) [reply src/main/java/org/gitlab4j/api/TopicsApi.java:160]: In the TopicsApi Documentation it says "To remove a topic avatar, use a blank value for the avatar attribute." 

Im not 100% sure how all of your Code Conventions work, but maybe does this work?

`        Response response = post(Response.Status.OK,new GitLabApiForm().withParam("avatar",""),"topics",id);
`

Comment (jmini) [reply src/main/java/org/gitlab4j/api/TopicsApi.java:160]: I have pushed 1cb014c7931bc6444c47cd4fef6a5d39de56a6d4

File: src/main/java/org/gitlab4j/api/models/ProjectFilter.java (hunk 4 above)
Comment (jmini) [reply src/main/java/org/gitlab4j/api/models/ProjectFilter.java:40]: Solved with 1cb36853bd56316acf99b6b7fbf9f48e4c094a5a

File: src/main/java/org/gitlab4j/api/models/Topic.java (hunk 3 above)
Comment (jmini) [reply src/main/java/org/gitlab4j/api/models/Topic.java]: Solved with 1cb36853bd56316acf99b6b7fbf9f48e4c094a5a

File: src/main/java/org/gitlab4j/api/models/Topic.java (hunk 2 above)
Comment (jmini) [reply src/main/java/org/gitlab4j/api/models/Topic.java]: Solved with 1cb36853bd56316acf99b6b7fbf9f48e4c094a5a

File: src/main/java/org/gitlab4j/api/models/Topic.java (hunk 1 above)
Comment (jmini) [reply src/main/java/org/gitlab4j/api/models/Topic.java]: Solved with 1cb36853bd56316acf99b6b7fbf9f48e4c094a5a

File: src/main/java/org/gitlab4j/api/models/Project.java (hunk 5 above)
Comment (jmini) [reply src/main/java/org/gitlab4j/api/models/Project.java:584]: Yes this was done in 1cb36853bd56316acf99b6b7fbf9f48e4c094a5a after my initial comment

**Comment by trp-samer-adra** [comment]
Any idea when this is getting released? Hoping to use it.

**Pull request merged**